File: matlab/tests/test_cqrrpt.m
function test_cqrrpt()
%test_cqrrpt  Correctness checks for randlapack.cqrrpt.
%
% Runs CQRRPT on tall matrices and checks:
%   1. Orthogonality of Q.
%   2. Factorization fidelity: A(:, J) ~= Q*R.
%   3. R is upper-trapezoidal.
%   4. J is a valid 1-based permutation.
%   5. Rank detection on a deliberately rank-deficient input.
%   6. Reproducibility from an explicit RNG state.
%   7. Error paths, including the tall-matrix requirement.
%
% Run from MATLAB after:
%   addpath('/path/to/randlapack-bindings/matlab')
%   test_cqrrpt

    types = {'double', 'single'};
    sizes = {[500, 50], [1000, 100], [2000, 200]};

    for ti = 1:numel(types)
        T = types{ti};
        for si = 1:numel(sizes)
            sz = sizes{si};
            A = cast(randn(sz(1), sz(2)), T);

            [Q, R, J] = randlapack.cqrrpt(A);

            % Full-rank input: the detected rank should be n.
            assert(size(R, 1) == sz(2), ...
                   '%s %dx%d: full-rank input gave rank %d, expected %d', ...
                   T, sz(1), sz(2), size(R, 1), sz(2));

            check_orthogonality(Q, T);
            check_factorization(A, Q, R, J, T);
            check_upper_trapezoidal(R, T);
            check_pivot_validity(J, sz(2));

            fprintf('PASS: %s %dx%d full rank\n', T, sz(1), sz(2));
        end

        % --- Rank-deficient input: the rank must be revealed exactly ---
        % Duplicated columns make the true rank exact and known, so this
        % assertion is tight on purpose. It is also a regression test for a
        % specific defect: CQRRPT passes J straight to LAPACK's geqp3, which
        % READS jpvt on entry, so a J buffer that is not zeroed makes the
        % pivoting depend on uninitialized memory. The visible symptom was
        % exactly this check -- the detected rank wandered between 1 and 40
        % across runs while the full-rank cases stayed green.
        m = 2000; r = 40; extra = 10;
        B = cast(randn(m, r), T);
        A = [B, B(:, 1:extra)];                 % rank r, r+extra columns
        A = A(:, randperm(size(A, 2)));         % scatter the duplicates

        [Q, R, J] = randlapack.cqrrpt(A);
        detected = size(R, 1);
        assert(detected == r, ...
               '%s: expected rank %d on a rank-deficient input, got %d', ...
               T, r, detected);
        check_orthogonality(Q, T);
        check_leading_factorization(A, Q, R, J, T);
        % With the rank fully revealed, the whole permuted matrix
        % reconstructs, not just the leading block.
        check_factorization(A, Q, R, J, T);
        fprintf('PASS: %s rank detection (%d of %d columns)\n', ...
                T, detected, size(A, 2));
    end

    % --- Reproducibility from an explicit state ---
    A  = randn(1000, 100);
    st = struct('counter', uint32([0 0 0 0]), 'key', uint32([0 11]));
    [Q1, R1, J1] = randlapack.cqrrpt(A, 'state', st);
    [Q2, R2, J2] = randlapack.cqrrpt(A, 'state', st);
    assert(isequal(Q1, Q2) && isequal(R1, R2) && isequal(J1, J2), ...
           'same explicit state must give bit-identical results');
    fprintf('PASS: reproducible from an explicit RNG state\n');

    [~, ~, ~, st_out] = randlapack.cqrrpt(A, 'state', st);
    assert(~isequal(st_out.counter, st.counter) || ~isequal(st_out.key, st.key), ...
           'returned state must differ from the input state');
    fprintf('PASS: RNG state advances\n');

    % --- d_factor is honoured ---
    [~, R_a, ~] = randlapack.cqrrpt(A, 'd_factor', 1.0);
    [~, R_b, ~] = randlapack.cqrrpt(A, 'd_factor', 2.0);
    assert(size(R_a, 1) == 100 && size(R_b, 1) == 100, ...
           'full-rank input must give rank 100 at either d_factor');
    fprintf('PASS: d_factor accepted across its range\n');

    % --- Error paths ---
    check_error(@() randlapack.cqrrpt(zeros(0, 5)), 'randlapack:cqrrpt:emptyInput');
    % Wide matrix: the tall requirement must be reported in MATLAB terms,
    % not as a dimension complaint from inside the sketching chain.
    check_error(@() randlapack.cqrrpt(randn(50, 500)), 'randlapack:cqrrpt:notTall');
    % Square is also too short once d_factor > 1.
    check_error(@() randlapack.cqrrpt(randn(100, 100)), 'randlapack:cqrrpt:notTall');
    check_error(@() randlapack.cqrrpt(A, 'd_factor', 0.5), ...
                'MATLAB:cqrrpt:notGreaterEqual');
    check_error(@() randlapack.cqrrpt(A, 'nosuchopt', 1), ...
                'randlapack:cqrrpt:unknownOption');
    check_error(@() randlapack.cqrrpt(int32(A)), 'MATLAB:cqrrpt:invalidType');
    fprintf('PASS: error paths\n');

    fprintf('test_cqrrpt: all checks passed\n');
end


function check_orthogonality(Q, T)
    [m, k] = size(Q);
    I   = cast(eye(k), T);
    err = norm(double(Q)' * double(Q) - double(I), 'fro');
    tol = sqrt(double(m * k)) * eps(T) * 100;
    assert(err < tol, ...
           'orthogonality check failed: ||Q^T Q - I||_F = %g > tol = %g', ...
           err, tol);
end


function check_factorization(A, Q, R, J, T)
    [m, n] = size(A);
    err = norm(double(A(:, J)) - double(Q) * double(R), 'fro') ...
          / norm(double(A), 'fro');
    tol = sqrt(double(m * n)) * eps(T) * 100;
    assert(err < tol, ...
           'factorization check failed: ||A(:,J) - QR||_F/||A||_F = %g > tol = %g', ...
           err, tol);
end


function check_leading_factorization(A, Q, R, J, T)
% The leading k pivoted columns are reproduced exactly whatever rank came
% back, because R(:, 1:k) is precisely their triangular factor. This is the
% invariant that survives a conservative rank estimate.
    [m, ~] = size(A);
    k = size(R, 1);
    lead = double(A(:, J(1:k)));
    err  = norm(lead - double(Q) * double(R(:, 1:k)), 'fro') ...
           / norm(double(A), 'fro');
    tol  = sqrt(double(m * k)) * eps(T) * 100;
    assert(err < tol, ...
           ['leading-block factorization failed: ' ...
            '||A(:,J(1:k)) - Q*R(:,1:k)||_F/||A||_F = %g > tol = %g'], ...
           err, tol);
end


function check_upper_trapezoidal(R, T)
    below = tril(double(R), -1);
    assert(norm(below, 'fro') == 0, ...
           'R must be upper-trapezoidal; found %g below the diagonal', ...
           norm(below, 'fro'));
    assert(isa(R, T), 'R must have the same class as A');
end


function check_pivot_validity(J, n)
    assert(numel(J) == n, 'J must have n entries');
    assert(isequal(sort(double(J(:)))', 1:n), ...
           'J must be a permutation of 1:n (1-based)');
end


function check_error(fn, expected_id)
    try
        fn();
    catch err
        assert(strcmp(err.identifier, expected_id), ...
               'expected error %s, got %s', expected_id, err.identifier);
        return;
    end
    error('expected error %s, but no error was raised', expected_id);
end

File: python/tests/test_cqrrpt.py
"""Correctness checks for randlapack.cqrrpt."""
import numpy as np
import pytest

import randlapack as rl


DTYPES = [np.float64, np.float32]


def tol_for(dtype, m, n):
    return np.sqrt(float(m * n)) * np.finfo(dtype).eps * 100


@pytest.mark.parametrize("dtype", DTYPES)
@pytest.mark.parametrize("shape", [(500, 50), (1000, 100), (2000, 200)])
def test_full_rank_factorization(dtype, shape):
    m, n = shape
    A = np.asfortranarray(
        np.random.default_rng(0).standard_normal(shape).astype(dtype))

    Q, R, J = rl.cqrrpt(A)

    # Full-rank input: the detected rank is n.
    assert R.shape[0] == n, f"expected rank {n}, got {R.shape[0]}"
    assert Q.shape == (m, n)
    assert J.shape == (n,)

    err = np.linalg.norm(A[:, J] - Q @ R) / np.linalg.norm(A)
    assert err < tol_for(dtype, m, n), f"||A[:,J] - QR||/||A|| = {err:g}"

    orth = np.linalg.norm(Q.T @ Q - np.eye(n, dtype=dtype))
    assert orth < tol_for(dtype, m, n), f"||Q^T Q - I|| = {orth:g}"


@pytest.mark.parametrize("dtype", DTYPES)
def test_pivots_are_zero_based_permutation(dtype):
    A = np.asfortranarray(
        np.random.default_rng(1).standard_normal((800, 60)).astype(dtype))
    _, _, J = rl.cqrrpt(A)
    # 0-based, matching scipy.linalg.qr(pivoting=True) -- not the 1-based
    # form CQRRPT uses natively and the MATLAB binding preserves.
    assert np.array_equal(np.sort(J), np.arange(60))


@pytest.mark.parametrize("dtype", DTYPES)
def test_rank_detection(dtype):
    """Duplicated columns give an exactly known rank.

    Also a regression test for the zero-initialization of J: CQRRPT passes it
    to LAPACK's geqp3, which reads jpvt on entry, so an uninitialized buffer
    makes the pivoting depend on heap garbage and the detected rank wander.
    """
    m, r, extra = 2000, 40, 10
    rng = np.random.default_rng(2)
    B = rng.standard_normal((m, r)).astype(dtype)
    A = np.asfortranarray(np.hstack([B, B[:, :extra]]))

    Q, R, J = rl.cqrrpt(A)

    assert R.shape[0] == r, f"expected rank {r}, got {R.shape[0]}"
    err = np.linalg.norm(A[:, J] - Q @ R) / np.linalg.norm(A)
    assert err < tol_for(dtype, m, r + extra)


@pytest.mark.parametrize("dtype", DTYPES)
def test_leading_block_identity(dtype):
    """A[:, J[:k]] == Q @ R[:, :k] holds whatever rank came back."""
    m, n = 1500, 80
    A = np.asfortranarray(
        np.random.default_rng(3).standard_normal((m, n)).astype(dtype))
    Q, R, J = rl.cqrrpt(A)
    k = R.shape[0]
    err = np.linalg.norm(A[:, J[:k]] - Q @ R[:, :k]) / np.linalg.norm(A)
    assert err < tol_for(dtype, m, k)


def test_c_order_input_matches_fortran_order():
    """The wrapper's asfortranarray coercion must not change the result."""
    A_c = np.random.default_rng(4).standard_normal((1000, 50))
    A_f = np.asfortranarray(A_c)
    assert not A_c.flags.f_contiguous

    Q1, R1, J1 = rl.cqrrpt(A_c, state=5)
    Q2, R2, J2 = rl.cqrrpt(A_f, state=5)

    assert np.array_equal(J1, J2)
    assert np.allclose(Q1, Q2, rtol=0, atol=0)
    assert np.allclose(R1, R2, rtol=0, atol=0)


def test_reproducible_from_explicit_state():
    A = np.asfortranarray(
        np.random.default_rng(6).standard_normal((1000, 50)))
    st = {"counter": np.zeros(4, dtype=np.uint32),
          "key": np.array([0, 11], dtype=np.uint32)}
    Q1, R1, J1 = rl.cqrrpt(A, state=st)
    Q2, R2, J2 = rl.cqrrpt(A, state=st)
    assert np.array_equal(Q1, Q2)
    assert np.array_equal(R1, R2)
    assert np.array_equal(J1, J2)


def test_state_advances():
    A = np.asfortranarray(
        np.random.default_rng(7).standard_normal((1000, 50)))
    st = {"counter": np.zeros(4, dtype=np.uint32),
          "key": np.array([0, 11], dtype=np.uint32)}
    _, _, _, st_out = rl.cqrrpt(A, state=st, return_state=True)
    assert not (np.array_equal(st_out["counter"], st["counter"])
                and np.array_equal(st_out["key"], st["key"]))


def test_wide_matrix_rejected_with_a_useful_message():
    A = np.random.default_rng(8).standard_normal((50, 500))
    with pytest.raises(ValueError, match=r"m >= d_factor\*n"):
        rl.cqrrpt(A)


def test_square_matrix_rejected():
    # Square is already too short once d_factor > 1.
    A = np.random.default_rng(9).standard_normal((100, 100))
    with pytest.raises(ValueError, match=r"m >= d_factor\*n"):
        rl.cqrrpt(A)


def test_error_paths():
    A = np.random.default_rng(10).standard_normal((1000, 50))
    with pytest.raises(ValueError):
        rl.cqrrpt(np.zeros((0, 5)))
    with pytest.raises(ValueError):
        rl.cqrrpt(A, d_factor=0.5)
    with pytest.raises(ValueError):
        rl.cqrrpt(A, eps=-1.0)
    with pytest.raises(TypeError):
        rl.cqrrpt(A.astype(np.int32))
    with pytest.raises(ValueError):
        rl.cqrrpt(np.random.default_rng(11).standard_normal((10, 10, 10)))
